plugins/outputs/bigquery/bigquery.go

```diff
@@ -0,0 +1,248 @@
+	s = append(s, timeSchema)
+	r = append(r, m.Time())
+
+	s, r = tagsSchemaAndValues(m, s, r)
+	s, r = valuesSchemaAndValues(m, s, r)
+
+	return &bigquery.ValuesSaver{
+		Schema: s.Relax(),
+		Row:    r,
+	}
+}
+
+func timeStampFieldSchema() *bigquery.FieldSchema {
+	return &bigquery.FieldSchema{
+		Name: timeStampFieldName,
+		Type: bigquery.TimestampFieldType,
+	}
+}
+
+func tagsSchemaAndValues(m telegraf.Metric, s bigquery.Schema, r []bigquery.Value) ([]*bigquery.FieldSchema, []bigquery.Value) {
+	for _, t := range m.TagList() {
+		s = append(s, tagFieldSchema(t))
+		r = append(r, t.Value)
+	}
+
+	return s, r
+}
+
+func tagFieldSchema(t *telegraf.Tag) *bigquery.FieldSchema {
+	return &bigquery.FieldSchema{
+		Name: t.Key,
+		Type: bigquery.StringFieldType,
+	}
+}
+
+func valuesSchemaAndValues(m telegraf.Metric, s bigquery.Schema, r []bigquery.Value) ([]*bigquery.FieldSchema, []bigquery.Value) {
+	for _, f := range m.FieldList() {
+		s = append(s, valuesSchema(f))
+		r = append(r, f.Value)
+	}
+
+	return s, r
+}
+
+func valuesSchema(f *telegraf.Field) *bigquery.FieldSchema {
+	return &bigquery.FieldSchema{
+		Name: f.Key,
+		Type: valueToBqType(f.Value),
+	}
+}
+
+func valueToBqType(v interface{}) bigquery.FieldType {
+	switch reflect.ValueOf(v).Kind() {
+	case reflect.Int, reflect.Int16, reflect.Int32, reflect.Int64:
+		return bigquery.IntegerFieldType
+	case reflect.Float32, reflect.Float64:
+		return bigquery.FloatFieldType
+	case reflect.Bool:
+		return bigquery.BooleanFieldType
+	default:
+		return bigquery.StringFieldType
+	}
+}
+
+func (b *BigQuery) insertToTable(metricName string, metrics []bigquery.ValueSaver) {
+	ctx := context.Background()
+	ctx, cancel := context.WithTimeout(ctx, b.Timeout.Duration)
+	defer cancel()
+
+	tableName := b.metricToTable(metricName)
+	table := b.client.DatasetInProject(b.Project, b.Dataset).Table(tableName)
+	inserter := table.Inserter()
+
+	if err := inserter.Put(ctx, metrics); err != nil {
+		b.Log.Errorf("inserting metric %q failed: %v", metricName, err)
+	}
+}
+
+func (b *BigQuery) metricToTable(metricName string) string {
+	if strings.Contains(metricName, "-") {
+
+		dhm := strings.ReplaceAll(metricName, "-", b.ReplaceHyphenTo)
+
+		if warned, _ := b.warnedOnHyphens[metricName]; !warned {
+			b.Log.Warn("Metric %q contains hyphens please consider using the rename processor plugin, falling back to %q", dhm)
+			b.warnedOnHyphens[metricName] = true
+		}
+
+		return dhm
+	}
+
+	return metricName
+}
+
+func (b *BigQuery) tableForMetric(metricName string) *bigquery.Table {
+	return b.client.DatasetInProject(b.Project, b.Dataset).Table(metricName)
+}
+
+// Close will terminate the session to the backend, returning error if an issue arises.
+func (b *BigQuery) Close() error {
+	return b.client.Close()
+}
+
+func init() {
+	outputs.Add("bigquery", func() telegraf.Output {
+		return &BigQuery{
+			Timeout:         defaultTimeout,
+			warnedOnHyphens: make(map[string]bool),
```

Comment:
Please do this in `Init()`.

Reply:
I moved this above. Did you mean doing the map initialization on the Connect()?

Reply:
I mean moving `b.warnOnHyphens = make(map[string]bool)` to `Connect()` instead of doing it here. Sorry for the confusion, `Init()` is the `Connect()` in input plugins... :-)